### PR TITLE
Redesign of data management, for a more modular approach.

### PR DESCRIPTION
**Pull request by richardjonker2000**

### CODE REVIEW

**Comment by T-Almeida** [comment]
Sry for the delay, I will start with the review but from the start in the usage example it is missing the imports

File: src/data.py [hunk 1]
```diff
@@ -8,258 +8,372 @@
...
+# stream_handler.setFormatter(formatter)
+# logger.addHandler(stream_handler)
 
-    for _, row in df.iterrows():
-        docs[row["filename"]].append({k:row[k] for k in ["ann_id", "label", "start_span", "end_span", "text"]})
-    
+
+class Corpus:
+    #  A class that represent a corpus. If you want a specific dataset, you need to implement a function that returns a corpus
+    def __init__(self, data: list ):
```
Comment (T-Almeida) [inline src/data.py]: Explain the format of the data argument should be in the readme, for me it seems the main entry point to add other datasets. If your data fits the that format it will always work right?

File: evaluation/eval.out
Comment (T-Almeida) [inline evaluation/eval.out:1]: This should have not been uploaded

File: README.md [hunk 2]
```diff
@@ -97,6 +97,106 @@ The evaluation (NER and entity linking) can be run in the [evaluation/](evaluati
 `python3  evaluation.py  train/test  PREDICTIONS_FILE.tsv`
 
 
+
+
+## How to use the architecture on a new Dataset
```
Comment (T-Almeida) [inline README.md]: Start instead with the standalone example, specifying the file

File: README.md [hunk 3]
```diff
@@ -97,6 +97,106 @@ The evaluation (NER and entity linking) can be run in the [evaluation/](evaluati
...
+# First, create a generic Corpus
+spanishCorpus = Spanish_Biomedical_NER_Corpus(
+    "../dataset/merged_data_subtask1_train.tsv", 
+    "../dataset/documents"
+)
+
+# Create a Corpus PreProcessor, which handles certain preprocessing tasks: 
+# merging annotations, filtering labels, and splitting the data.
+spanishCorpusProcessor = CorpusPreProcessor(spanishCorpus)
+spanishCorpusProcessor.merge_annoatation()
```
Comment (T-Almeida) [inline README.md]: This function is optional? It does not have arguments, how the user knows if he should run or not?

If it is something that we should always run, then it should not be a public function. Instead, we should just run it inside, even if it does nothing for some datasets.

File: README.md [hunk 4]
```diff
@@ -97,6 +97,106 @@ The evaluation (NER and entity linking) can be run in the [evaluation/](evaluati
...
+spanishCorpusProcessor.filter_labels(classes)
+
+# Split the corpus into training and testing sets with a 33% split.
+train_corpus, test_corpus = spanishCorpusProcessor.split_data(0.33)
+
+# Create a CorpusTokenizer, using the CorpusPreProcessor. 
+# This internally tokenizes the dataset and splits the documents.
+tokenized_train_corpus = CorpusTokenizer(train_corpus, tokenizer, CONTEXT_SIZE)
```
Comment (T-Almeida) [inline README.md]: CorpusTokenizer cannot be handled internally as well? Why it needs to be the end user to handling this? 

We can just pass to the CorpusDataset the corpus, normal tokenizer and context size, no? What is advantage to have it like this?

File: src/data.py [hunk 5]
```diff
@@ -8,258 +8,372 @@
...
 
-def load_train_val( tokenizer,
-                    context_size=64,
-                    test_split_percentage=0.15,
-                    train_transformations=None,
-                    train_augmentations=None,
-                    entity= ['SYMPTOM', 'PROCEDURE', 'DISEASE', 'PROTEIN', 'CHEMICAL'],
-                    test_transformations=None
-                    ):
+# import logging
```
Comment (T-Almeida) [inline src/data.py:16]: Adding a logger would be a nice idea

File: src/data.py [hunk 6]
```diff
@@ -8,258 +8,372 @@
...
+            print(f"Warning, we found {total_collisions} collisions that were automaticly handle by merging strategy")
+        
+    def split_data(self, test_split_percentage):
+        trainCorpus, testCorpus = self.corpus.split(test_split_percentage)
+        return  CorpusPreProcessor(trainCorpus), CorpusPreProcessor(testCorpus)
+    
 
 
-class DocumentReaderDatasetForTraining(torch.utils.data.Dataset):
+def Spanish_Biomedical_NER_Corpus(file_path, documents_folder):
```
Comment (T-Almeida) [inline src/data.py]: It's okay, but it would not make more sense to be a class that extends from Corpus? What was the issue here?

File: src/hf_trainer.py
Comment (T-Almeida) [inline src/hf_trainer.py:1]: I run it and it still works

File: src/inference.py
Comment (T-Almeida) [inline src/inference.py:1]: I run it and it still works

**Review by T-Almeida — COMMENTED** [review]
Looks good, some question about some design choices and README needs a bit more care to become more clear. Focus should be on how to quickly use other datasets (what I need to change and to which format, with small exemples)

File: src/data.py (hunk 1 above)
Comment (richardjonker2000) [reply src/data.py]: line 36: #Data needs to be formated as follows {"doc_id":id, "text": doucment_text, "annotations":[{'label': LABEL, 'start_span': num, 'end_span': num}, ... ]  }   
I agree should be in Readme.  But yes if the data is in the same format all other funcitons work.

File: README.md (hunk 3 above)
Comment (richardjonker2000) [reply README.md]: The function is optional. It merges overlapping annotaitons, it does not need any arguments.  

Not sure if the remainder code will support overlapping entities, but the function is not mandatory to run. Highly advised to run it.

File: README.md (hunk 4 above)
Comment (richardjonker2000) [reply README.md]: It can be handled internally but its also for development its more clearer what is being done. 

The internal data structure also changes, which is why it is a seperate class. 

We can think about merging but we shoud discuss this more first

**Comment by richardjonker2000** [comment]
To resolve

- [ ] readme updates
- [ ] logger
- [ ] Spanish_Biomedical_NER_Corpus - as a funtion vs as a class
- [ ] CorpusTokenizer, as an extension of another class
- [ ] PreProcessor: merge_annoatation
- [ ] clean files

File: src/data.py (hunk 6 above)
Comment (T-Almeida) [reply src/data.py]: Indentation might be wrong

```python
class Spanish_Biomedical_NER_Corpus(Corpus):
	def __init__(file_tsv_file, docs_folder):
		annotations = defaultdict(list)
		df = pd.read_csv(file_path, sep="\t")
		for _, row in df.iterrows():
			annotations[row["filename"]].append({k:row[k] for k in ["label", "start_span", "end_span"]})

		#load the documents
		document_text = {}
		for file in annotations.keys():
			with open(os.path.join(documents_folder,f"{file}.txt"), 'r') as f:
				document_text[file] = ''.join([line for line in f]).strip()

		data = [{"doc_id":k, "text":document_text[k], "annotations":v} for k,v in annotations.items()]   

		super().__init__(data)

```

**Review by T-Almeida — APPROVED** [review]
Readme seems good, and Richard said it was tested
